engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniPassTraverser.java

```diff
@@ -0,0 +1,153 @@
+   */
+  public static Module compileModuleWithMiniPass(Module moduleIr, MiniIRPass miniPass) {
+    var preparedMiniPass = miniPass.prepareForModule(moduleIr);
+    logPrepare(moduleIr, miniPass, preparedMiniPass);
+    var newChildren = new ArrayList<Expression>();
+    boolean[] childrenChanged = {false};
+    Module newModuleIr;
+    moduleIr.mapExpressions(
+        (ch) -> {
+          var newChild = compileRecursively(ch, preparedMiniPass);
+          newChildren.add(newChild);
+          if (ch != newChild) {
+            childrenChanged[0] = true;
+          }
+          return newChild;
+        });
+    if (childrenChanged[0]) {
+      var index = new int[1];
+      newModuleIr = moduleIr.mapExpressions((old) -> newChildren.get(index[0]++));
+    } else {
+      newModuleIr = moduleIr;
+    }
+    return miniPass.transformModule(newModuleIr);
+  }
+
+  public static Expression compileInlineWithMiniPass(Expression exprIr, MiniIRPass miniPass) {
+    var preparedMiniPass = miniPass.prepare(exprIr);
+    var newIr = compileRecursively(exprIr, preparedMiniPass);
+    if (!preparedMiniPass.checkPostCondition(newIr)) {
+      throw new CompilerError("Post condition failed after applying mini pass " + preparedMiniPass);
+    }
+    return newIr;
+  }
+
+  private static Expression compileRecursively(Expression expr, MiniIRPass miniPass) {
+    var preparedMiniPass = miniPass.prepare(expr);
+    logPrepare(expr, miniPass, preparedMiniPass);
+    Expression newIr;
+    var childExpressions = new ArrayList<Expression>();
+    expr.mapExpressions(
+        (ch) -> {
+          childExpressions.add(ch);
+          return ch;
+        });
+    if (childExpressions.isEmpty()) {
+      newIr = expr;
+    } else {
+      var changed = false;
+      for (var i = 0; i < childExpressions.size(); i++) {
+        var child = childExpressions.get(i);
+        var newChild = compileRecursively(child, preparedMiniPass);
+        if (child == newChild) {
+          continue;
+        }
+        changed = true;
+        if (!preparedMiniPass.checkPostCondition(newChild)) {
+          throw new CompilerError(
+              "Post condition failed after applying mini pass " + preparedMiniPass);
+        }
+        childExpressions.set(i, newChild);
+      }
+      if (changed) {
+        var index = new int[1];
+        newIr = expr.mapExpressions((old) -> childExpressions.get(index[0]++));
+      } else {
+        newIr = expr;
+      }
+    }
+    var transformedIr = preparedMiniPass.transformExpression(newIr);
+    logTransform(newIr, preparedMiniPass, transformedIr);
+    if (!preparedMiniPass.checkPostCondition(transformedIr)) {
+      throw new CompilerError("Post condition failed after applying mini pass " + preparedMiniPass);
+    }
+    return transformedIr;
+  }
+
+  private static void logPrepare(IR ir, MiniIRPass pass, MiniIRPass preparedPass) {
+    if (!logger.isTraceEnabled()) {
+      return;
+    }
+    var irName = irToSimpleString(ir);
```

Comment:
I see a lot of calls into `irToSimpleString` while doing `sbt buildEngineDistribution`:
![obrazek](https://github.com/user-attachments/assets/efbfe90e-5bec-42b5-9d99-7bbfa8f91ee5)
and they are slowing down the whole _" Generating index for ..."_ quite visibly.

Reply:
I have removed the excessive use of `logger` in subsequent PRs. Moreover I believe we should move towards **non-recursive** iteration like:
```diff
diff --git engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniIRPass.java engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniIRPass.java
index 9dc9685f1e..9e5fb59fef 100644
--- engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniIRPass.java
+++ engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniIRPass.java
@@ -132,7 +132,7 @@ public abstract class MiniIRPass implements ProcessingPass {
    * @return the transformed IR
    */
   public static <T extends IR> T compile(Class<T> irType, T ir, MiniIRPass miniPass) {
-    var newIr = MiniPassTraverser.compileRecursively(ir, miniPass);
+    var newIr = MiniPassTraverser.compileDeep(ir, miniPass);
     assert irType.isInstance(newIr)
         : "Expected "
             + irType.getName()
diff --git engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniPassTraverser.java engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniPassTraverser.java
index 2157dfb7ad..f34091a7d5 100644
--- engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniPassTraverser.java
+++ engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniPassTraverser.java
@@ -1,54 +1,98 @@
 package org.enso.compiler.pass;
 
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
 import org.enso.compiler.core.IR;
 import org.enso.compiler.core.ir.Expression;
 import org.enso.compiler.core.ir.Module;
 
 /** Implementation of {@link MiniIRPass#compile}. */
 final class MiniPassTraverser {
-  private MiniPassTraverser() {}
-
-  static IR compileRecursively(IR ir, MiniIRPass miniPass) {
-    var preparedMiniPass =
-        switch (ir) {
-          case Module m -> miniPass.prepareForModule(m);
-          case Expression e -> miniPass.prepare(e);
-          default -> throw new IllegalArgumentException("" + ir);
-        };
-    IR newIr;
-    var childExpressions = new ArrayList<Expression>();
+  private final MiniIRPass miniPass;
+  private List<IR> in;
+  private final List<IR> out;
+  private int outIndex;
+
+  private MiniPassTraverser(MiniIRPass miniPass, List<IR> out, int outIndex) {
+    this.miniPass = miniPass;
+    this.out = out;
+    this.outIndex = outIndex;
+  }
+
+  private boolean enqueue(Collection<MiniPassTraverser> queue) {
+    if (in == null) {
+      var ir = out.get(outIndex);
+      in = enqueueSubExpressions(queue, ir, miniPass);
+      return !in.isEmpty();
+    } else {
+      return false;
+    }
+  }
+
+  private void convertExpression() {
+    if (outIndex != -1) {
+      var oldIr = out.get(outIndex);
+      var index = new int[1];
+      var newIr = oldIr.mapExpressions((old) -> (Expression) in.get(index[0]++));
+      var transformedIr =
+          switch (newIr) {
+            case Module m -> miniPass.transformModule(m);
+            case Expression e -> miniPass.transformExpression(e);
+            default -> throw new IllegalArgumentException("" + oldIr);
+          };
+      if (oldIr != transformedIr) {
+        out.set(outIndex, transformedIr);
+      }
+      outIndex = -1;
+    }
+  }
+
+  static IR compileDeep(IR root, MiniIRPass miniPass) {
+    var result = new IR[] {root};
+    var rootTask = new MiniPassTraverser(miniPass, Arrays.asList(result), 0);
+    var stackOfPendingIrs = new LinkedList<MiniPassTraverser>();
+    stackOfPendingIrs.add(rootTask);
+    while (!stackOfPendingIrs.isEmpty()) {
+      if (stackOfPendingIrs.peekLast().enqueue(stackOfPendingIrs)) {
+        // continue descent
+        continue;
+      }
+      var deepestIr = stackOfPendingIrs.removeLast();
+      deepestIr.convertExpression();
+    }
+    assert result[0] != null;
+    return result[0];
+  }
+
+  /**
+   * @param queue queue to put objects in
+   * @param ir IR to process
+   * @param miniPass process with this mini pass
+   * @return {@code true} if the has been modified with new tries to process first
+   */
+  private static List<IR> enqueueSubExpressions(
+      Collection<MiniPassTraverser> queue, IR ir, MiniIRPass miniPass) {
+    var childExpressions = new ArrayList<IR>();
     ir.mapExpressions(
         (ch) -> {
           childExpressions.add(ch);
           return ch;
         });
-    if (childExpressions.isEmpty()) {
-      newIr = ir;
-    } else {
-      var changed = false;
+    if (!childExpressions.isEmpty()) {
+      var preparedMiniPass =
+          switch (ir) {
+            case Module m -> miniPass.prepareForModule(m);
+            case Expression e -> miniPass.prepare(e);
+            default -> throw new IllegalArgumentException("" + ir);
+          };
+
       for (var i = 0; i < childExpressions.size(); i++) {
-        var child = childExpressions.get(i);
-        var newChild = (Expression) compileRecursively(child, preparedMiniPass);
-        if (child == newChild) {
-          continue;
-        }
-        changed = true;
-        childExpressions.set(i, newChild);
-      }
-      if (changed) {
-        var index = new int[1];
-        newIr = ir.mapExpressions((old) -> childExpressions.get(index[0]++));
-      } else {
-        newIr = ir;
+        queue.add(new MiniPassTraverser(preparedMiniPass, childExpressions, i));
       }
     }
-    var transformedIr =
-        switch (newIr) {
-          case Module m -> miniPass.transformModule(m);
-          case Expression e -> miniPass.transformExpression(e);
-          default -> throw new IllegalArgumentException("" + newIr);
-        };
-    return transformedIr;
+    return childExpressions;
   }
 }
```